### ISSUE TITLE
Swoole\Coroutine\Http\Client Broken without Content-Length header

### ISSUE THREAD

**Issue by mgrooms** [swoole/swoole-src#1450]
Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

The Swoole\Coroutine\Http\Client works fine when contacting a server that responds with a Content-Length header. However, when contacting a server that responds with a Connection: close header but not a Content-Length header, the client does not interpret  the TCP connection being closed as end of the HTTP message transmission. Instead, it waits for the timeout expire and fails to return the message body to the caller.

2. What did you expect to see?

The HTTP protocol defines the Content-Length header for use when the server knows the length of the content in advance. When the server does not know the length ( usually dynamic content ), it can specify Connection: close to denote that the close of the TCP Connection is the end of the HTTP content. The Swoole http client should handle this correctly. I didn't have time to dig the swoole source code to find the root cause.

3. What did you see instead?

The Swoole client does not behave correctly when the TCP Connection is closed. It seems to ignore the socket being closed and waits for the timeout to expire. When reading data using the C socket API, a zero length message ( recv() call completes successfully with a read length of zero ) denotes that the socket has been closed by the remote host ( FIN packet has been sent ). My guess is that this scenario isn't handled correctly by swoole.

4. What version of Swoole are you using (`php --ri swoole`)?

swoole

swoole support => enabled
Version => 2.0.12
Author => tianfeng.han[email: mikan.tenny@gmail.com]
kqueue => enabled
spinlock => enabled
rwlock => enabled
async redis client => enabled
async http/websocket client => enabled
openssl => enabled
http2 => enabled
pcre => enabled
zlib => enabled
mutex_timedlock => enabled
pthread_barrier => enabled

Directive => Local Value => Master Value
swoole.aio_thread_num => 2 => 2
swoole.display_errors => On => On
swoole.use_namespace => On => On
swoole.fast_serialize => Off => Off
swoole.unixsock_buffer_size => 8388608 => 8388608

5. What is your machine environment used (including version of kernel & php & gcc) ? 

FreeBSD 10.3-RELEASE-p24
php71-7.1.12
pecl-swoole-2.0.12

6. If you are using ssl, what is your openssl version?

N/A

Here is an example using the swoole framework that should demonstrate the issue. Sadly, I couldn't get multiple listening sockets to work due to what looks like another bug. The port1 and port2 receive functions never get called :( Maybe a Swoole developer will have more luck. To run the test, just send a curl call to localhost like so ...

```

# test port 9101
$ curl http://localhost:9100/test1

# test port 9102
$ curl http://localhost:9100/test2

<?php

function test( $port )
{
        echo "test to port {$port} begin\n";
        $cli = new Swoole\Coroutine\Http\Client( '127.0.0.1', $port );
        $cli->set([
                'timeout' => 10
        ]);
        $cli->setHeaders([
                'Connection' => 'close',
                'Accept' => '*/*'
        ]);
        $cli->get( '/' );
        echo "response = '{$cli->body}'\n";
        echo "test to port {$port} complete\n";
}

$server = new swoole_http_server( '127.0.0.1', 9100, SWOOLE_BASE );

$port1 = $server->listen( '127.0.0.1', 9101, SWOOLE_SOCK_TCP );
$port1->on( 'receive', function( $server, $fd, $from_id, $recv ) {
        echo "test1 recv = '{$recv}'\n";
        //
        // send http result with content-length ( client completes )
        //
        $serv->send( $fd,
                "HTTP/1.1 200 OK\n".
                "Connection: close\n".
                "Content-Type: text/html\n".
                "Content-Length: 12\n".
                "\n".
                "test1 result"
         );
        $serv->close( $fd );
});

$port2 = $server->listen( '127.0.0.1', 9102, SWOOLE_SOCK_TCP );
$port2->on( 'receive', function( $server, $fd, $from_id, $recv ) {
        echo "test2 recv = '{$recv}'\n";
        //
        // send http result without content-length ( client stalls )
        //
        $serv->send( $fd,
                "HTTP/1.1 200 OK\n".
                "Connection: close\n".
                "Content-Type: text/html\n".
                "\n".
                "test2 result"
         );
        $serv->close( $fd );
});

$server->on( 'request', function( $req, $resp ) {
        switch( $req->server['request_uri'] )
        {
                case '/test1':
                        $resp->end( 'running test1' );
                        test( 9101 );
                        break;
                case '/test2':
                        $resp->end( 'running test2' );
                        test( 9102 );
                        break;
        }
});

$server->start();

?>

```

**Comment by mgrooms** [comment]
I was able to look at the swoole source code a bit and diagnose what I believe to be the issue. When the server transmits the HTTP response and closes the connection, the swoole http client receives an onClose event. However, the php_http_parser_execute is designed to be passed a zero length buffer to indicate the connection has been closed so it can complete processing of the http message. I have included a small patch that seems safe and fixes the issue for me, but I'm not sure if it's the 'right thing to do'. Hope this helps.

[swoole_http_client_coro.c.diff.txt](https://github.com/swoole/swoole-src/files/1632616/swoole_http_client_coro.c.diff.txt)
